### DISCUSSION TITLE
How does client-side layouting work for GPort?

### DISCUSSION THREAD

**Discussion by srcnuzn** [eclipse-glsp/glsp#1070] [Q&A]
For example, both ports in the `GCompartment` below are added overlappingly at the same position instead of being layouted vertically. Note that the vbox layouting works if I replace the ports with simple labels. Am I missing something?

```java
new GCompartmentBuilder()
    .layout(GConstants.Layout.VBOX)
    .layoutOptions(Map.of(
            GLayoutOptions.KEY_H_ALIGN, GConstants.HAlign.LEFT,
            GLayoutOptions.KEY_V_ALIGN, GConstants.VAlign.CENTER))
    .add(new GPortBuilder(DefaultTypes.PORT)
            .add(new GLabelBuilder().text("PORT_1").build())
            .build())
    .add(new GPortBuilder(DefaultTypes.PORT)
            .add(new GLabelBuilder().text("PORT_2").build())
            .build())
    .build()
```

<img src="https://github.com/eclipse-glsp/glsp/assets/61967763/30e20107-72fe-4d6c-a2dd-d0a8238accc4" width="30%" height="30%">

**Comment by CamilleLetavernier** [comment]
Hi,

By default, I believe GPort is a standard Node (Although it depends on how you map it in your own application). For this reason, I would expect it to behave like any other child node.

In your case, it seems that the Port doesn't have a size, or its size is not properly taken into account by the parent (Vbox) Layout. This could be because you didn't declare a Layout for your Port, so the parent layout doesn't know what to do with it.

Can you try setting a `vbox` or `hbox` layout on your Ports?

**Reply by srcnuzn** [reply]
Hey,

unfortunately, the `GPortBuilder` does not offer to set layout. I'm not sure if `GPort ` is layoutable at all, as it does not extend `GLayouting` or `GEdgeLayoutable` (is this intended?). Also, setting the size of the ports had no effect on the layouting (but the size change was visible).

What helped was your suggestion from my other question, namely to add a surrounding `GCompartment ` (see [here](https://github.com/eclipse-glsp/glsp/discussions/1069#discussioncomment-6526574)).

With this code the layouting works, but is it the intended way for client-side layouting of ports?

```java
new GCompartmentBuilder()
    .layout(GConstants.Layout.VBOX)
    .layoutOptions(Map.of(
            GLayoutOptions.KEY_H_ALIGN, GConstants.HAlign.LEFT,
            GLayoutOptions.KEY_V_ALIGN, GConstants.VAlign.CENTER))
    .add(new GCompartmentBuilder()
            .add(new GPortBuilder()
                    .add(new GLabelBuilder().text("PORT_1").build())
                    .build())
            .build())
    .add(new GCompartmentBuilder()
            .add(new GPortBuilder()
                    .add(new GLabelBuilder().text("PORT_2").build())
                    .build())
            .build())
    .build()
```
<img src="https://github.com/eclipse-glsp/glsp/assets/61967763/1759eb62-7a24-4bb4-88cd-1bf70ac69718" width="30%" height="30%">
